### ISSUE TITLE
resolve vulnerabilities

### ISSUE THREAD

**Issue by mihai-sysbio** [MetabolicAtlas/MetabolicAtlas#719]
Aim for no warnings https://github.com/MetabolicAtlas/MetabolicAtlas/security/dependabot

**Comment by e0** [comment]
If the dependabot issue is not directly related to the versions, it might be good to not change any versions for now. For example it seems like node 14 is needed for now and the latest Docker version is causing some problems for deployment.

**Comment by mihai-sysbio** [comment]
The dependabot warnings are afaik only for `package.json`, not related with Docker.

**Comment by e0** [comment]
Thanks for the reply @mihai-sysbio . Looking at https://github.com/MetabolicAtlas/MetabolicAtlas/security/code-scanning, I'm not sure exactly how `package.json` is related to these issues. Can you please provide more details?

**Comment by mihai-sysbio** [comment]
There is a different issue for the CodeQL warning. This issue is different - the OP links to dependabot.

**Comment by e0** [comment]
Ah that's right, sorry. That link is not working though.

**Comment by mihai-sysbio** [comment]
It does, as long as one is signed in.

**Comment by e0** [comment]
I have double checked that I am signed in. Is it an issue with user permissions?

**Comment by mihai-sysbio** [comment]
You're right, it was a permission issue. Somewhere in the org settings under `Security & analysis`. It should be fixed now.

**Comment by nanjiangshu** [comment]
Now I see the same page as yours @mihai-sysbio . Thanks for fixing the permission.

**Comment by e0** [comment]
Relevant issue for `ansi-html`: https://github.com/Tjatse/ansi-html/issues/21

**Comment by inghylt** [comment]
https://docs.github.com/en/code-security/supply-chain-security/managing-vulnerabilities-in-your-projects-dependencies/about-dependabot-security-updates

I think it would make sense to try this out for updating the dependencies. And maybe keep it so that it is easier to update in the future since it will automatically create a PR that we can review an see that everything is working accordingly. That would avoid us having to do this type of update where a lot of dependencies are changed at the same time

https://docs.github.com/en/code-security/supply-chain-security/managing-vulnerabilities-in-your-projects-dependencies/configuring-dependabot-security-updates#enabling-or-disabling-dependabot-security-updates-for-an-individual-repository
This is how we add it in that case, however I do not have permissions to do so

**Comment by mihai-sysbio** [comment]
We have had this set up before, and turned it off. Past experience has shown that package updates can break the functionality of the website, so updates have to be tested. Therefore, it felt easier to wait a bit and do a batch update, followed by manual testing until we have E2E tests in place. I'm happy to revisit this decision though, but that is not really helping advance the current issue. How about having a manual update and testing now, and create issues for managing this better in the future?

**Comment by e0** [comment]
I think it would help both for this issue and in the future to set this up now. The bot creates the PR with the necessary updates but we should still test the branch manually before merging. In the future it would be a lot easier to review single package updates instead of doing it in batch. For now it might be more work in case anything breaks from the bot PR but should still facilitate speeding things up.

**Comment by mihai-sysbio** [comment]
>  In the future it would be a lot easier to review single package updates instead of doing it in batch. 

I'm not really sure I follow, perhaps we're imagining the package update to be done differently.

The way I see it, updating the packages is the easy and quick part, since the only thing needed is to remove the lock files and build the stack. I'm counting on our dependencies to be updated, rather than us adopting the new versions eagerly.

Checking all the functionality of the website can be tedious. If the branch still needs to be manually tested, that's the time consuming bit that we cannot afford to do too often. There are 13 alerts since June (a bit in batch, see timestamps). Doing a manual test of the website more or less every sprint feels unacceptable to me; I'd rather have that time invested in E2E tests that we can benefit from on all PRs.

**Comment by e0** [comment]
> Doing a manual test of the website more or less every sprint feels unacceptable to me

The benefit of testing a single package update is that it's much easier to track down the problem. If we are updating 10 packages and notice a problem it might take a while to identify which package caused it. Also I think the testing could be done more specific to the package if possible, so that the entire website does not need to be manually tested each time. 

If we don't want to do this too frequently, we can just let the PRs sit there until we have time to go through them. I don't see any downside of adding the bot and the advantage is that it streamlines our workflow.

**Comment by inghylt** [comment]
If we have a PR with a fix automatically created each time there is a vulnerability reported I think that would lead to:
- us not having vulnerable dependencies (multiple marked with high severity) waiting to be fix for more than 6 months (!)
- that it would not take time to check that functionality is kept in place, since it will likely only be one package updated at a time. It will be easy to see which part of the page will be affected. And as @e0 says, we don't have to review the PR as soon as it is created, we can just wait until there is time

I think we should discuss this live at standup today as well

**Comment by mihai-sysbio** [comment]
Some notes here for the discussion:
- difference between vulnerabilities in code vs dependencies
- testing of forced updates of deep dependencies
- update policy in `package.json`: `^` not always sufficient, not always used

**Comment by e0** [comment]
Thanks again for the nice discussion everyone! It was good to get clarity on some details. Don't think I mentioned it during the call but if this (trying to update all packages) is something we do only once a year, then I fully support it. Should be a good thing to go through the packages we are using even just to learn about the newer updates. It also makes sense that testing after this kind of changes should be done more thoroughly compared to if we are just changing a single package.

**Comment by inghylt** [comment]
Yes, very nice meeting!

They suggested way forward as I see it is to have an iterative process, since we don't know if there will be problems or not. Some steps on the way could be:

- Trying to update all dependencies to have package.json in a good state and test thoroughly 
  - If that will create broken functionality, a good way forward would be to try to only update dependencies related to the warnings and/or check dependencies which have major version changes individually
- Once the work with updating package.json and validating the changes is done, it makes sense to merge
- Once the merge is finished, enable Dependabot PR:s (but not automatically merging them). If there are warnings left, look into the suggested PR by Dependabot
- If we feel like automatic Dependabot PR:s are taking too much time in the future, we can just disable it again

**Comment by inghylt** [comment]
yarn upgrade-interactive --latest

This is a nice command to use

Before this issue is considered done, a small update to the docs with a "brain-dump" should be added since we probably want to do this again but in a distant future

**Comment by e0** [comment]
When trying out different packages for this type of work, it could be convenient to temporarily mount the `package.json` by adding the following line in `docker-compose-local.yml`. After adding it, run `start-stack` again and the change should be reflected and any local changes to `package.json` should be synced to the docker container.

![image.png](https://images.zenhubusercontent.com/60992f121594d60ae1132a98/3839fa4e-9050-4b8b-a206-1ab7068cf688)

**Comment by inghylt** [comment]
Summary 1, updating package.json

For some reason, and this was verified when pair programming with @e0 , adding `- ./frontend/package.json:/project/package.json` to docker-compose-local.yml does not make changes take effect immediately in the running frontend container. I had to do `stop-stack` and then `start-stack` for package.json to change. So my (suboptimal)  workflow tended to be

1. Make changes to package.json
2. `start-stack` 
3. `ma-exec frontend cat package.json` (if I want to be sure that the updated package.json is in place)
4. `ma-exec frontend yarn install`
5. `stop-stack`
6. `start-stack` (this is required for making the downloaded node_modules in step 4 t actually be used)
7. see if something is broken

If the version update caused the frontend container to break entirely and not run, `logs frontend` was used to identify what might be wrong. Change package.json and then run `build-stack` again to have a running stack again

**Comment by inghylt** [comment]
Summary 2:

All dependencies in frontend/package.json have been updated to latest possible version without breaking functionality. Efforts were made, both by me and with the support of @e0 and @nanjiangshu, to be able to update node-sass and sass-loader by updating frontend docker image from node 14 to node 16 and 17 and trying out different versions of the dependencies etc. Unfortunately when one problem was solved a new one kept popping up. Issue #836 was created since this solving this seems to not be entirely straightforward

**Comment by e0** [comment]
Great job @inghylt! Really nice summaries and great work creating the new issue that details all of the errors. Should make it really easy to pick up the work again for that issue. 👏

**Comment by mihai-sysbio** [comment]
I might have stumbled on a fix for #836, but that issue only deals with the `frontend` container. Node v14 is used also for the `api` container, but a bump up there might require a `webpack` bump. However, I have noticed two vulnerabilities reported by GitHub for `api` that might be resolved with @inghylt's workflow in the posts above. Perhaps it makes sense to leave all the `api` problems for a separate issue?

**Comment by nanjiangshu** [comment]
Thanks for the great summary @inghylt has made. I'm not quite sure if `api/package.json` has been touched already. If not, maybe it's the best moment to work with the `API` now while the experience from the `frontend` is still fresh. If we still encounter the same `node-sass` problem for the API, we can then leave that problem to the issue #836 as well.

**Comment by inghylt** [comment]
Summary to close this issue:
Turns out that 6/8 of the remaining vulnerabilities, i.e. all warnings expect for ansi-regex, come from package vue-cli, which we already have the latest version of. I have created an issue https://github.com/vuejs/vue-cli/issues/6976 for fixing the severe ones. Since these vulnerabilities is not a dependency of ours and since we can't really update anything as this is due to vue-cli, I suggest we dismiss all the related vulnerabilities and see what will happen to the issue I have created and if there will be any fixes in a later vue-cli release. For ansi-regex, this is also not a dependency of ours but a nested one. Since we can't use Dependabots automatic update, and since we have created issue #856 for sanitizing user input ([see related discussion](https://github.com/chalk/ansi-regex/commit/8d1d7cdb586269882c4bdc1b7325d0c58c8f76f9)), I suggest we dimsiss this vulnerability for now as well. [There seem to be a Dependabot bug related to this one as well](https://github.com/dependabot/dependabot-core/issues/4222).

For future, I think it makes sense to continue to keep an eye on the Dependabot alerts and fix when possible/needed, prioritizing High severity vulnerabilities. This work should have a time cap though. We should also make sure that fixing code scanning alerts is not underprioritized compared to this

**Comment by inghylt** [comment]
I have now dismissed the vulnerabilities as suggested here and at Slack. Copied the information from Dependabot if we want to have information about the dismissed vulnerabilities in the future
[2022-02-03-dismiss-vulnerabilities.md](https://github.com/MetabolicAtlas/MetabolicAtlas/files/7993579/2022-02-03-dismiss-vulnerabilities.md)

**Comment by inghylt** [comment]
Added section Dependency management under Development in the internal wiki to document knowledge obtained from working with this issue. Feel free to add relevant information there

**Comment by inghylt** [comment]
Bumped to a five
